### PR TITLE
Add dark mode support to welcome theme and unify implementation

### LINKED ISSUE #26178: Support dark mode, at least for the login pages
### Description

A lot of apps support dark mode. We also support it. Our users select dark mode both on the website and in their user panel. However, when they are logged out, they are redirected to Keycloak's login page, and it bothers their eyes. Especially in the middle of the night.

### Discussion

_No response_

### Motivation

We want to deliver a unified UX for our users. We want them to be able to experience a dark theme out of the box for the login page too.

### Details

I know it's possible for us to create a custom theme. But that's a huge overhead. I think a default dark mode should be available out of the box, and it should be activated using a URL parameter like `&dark=true`.

### PR DESCRIPTION
**Pull request by jonkoops**
Closes #26178

### CODE REVIEW

**Review by agagancarczyk — DISMISSED** [review]

**Review by ssilvert — DISMISSED** [review]
@jonkoops LGTM once the tests pass.

File: themes/src/main/resources/theme/keycloak.v2/login/template.ftl [hunk 1]
```diff
@@ -6,13 +6,33 @@
...
     <meta name="robots" content="noindex, nofollow">
+    <meta name="color-scheme" content="light dark">
 
     <#if properties.meta?has_content>
         <#list properties.meta?split(' ') as meta>
             <meta name="${meta?split('==')[0]}" content="${meta?split('==')[1]}"/>
         </#list>
     </#if>
     <title>${msg("loginTitle",(realm.displayName!''))}</title>
+    <script type="module" async blocking="render">
```
Comment (jonkoops) [inline themes/src/main/resources/theme/keycloak.v2/login/template.ftl:17]: This script tag should be rendered based on whether dark mode is enabled or not. If has no purpose if dark mode support is not enabled.

File: themes/src/main/resources/theme/keycloak.v2/login/theme.properties [hunk 2]
```diff
@@ -87,4 +87,7 @@ kcLoginOTPListItemHeaderClass=pf-v5-c-tile__header pf-m-stacked
 kcLoginOTPListItemIconBodyClass=pf-v5-c-tile__icon
 kcLoginOTPListItemIconClass=fa fa-mobile
 kcLoginOTPListItemTitleClass=pf-v5-c-tile__title
-kcLoginOTPListSelectedClass=pf-m-selected
\ No newline at end of file
+kcLoginOTPListSelectedClass=pf-m-selected
+
+kcDarkModeClass=pf-v5-theme-dark
+kcDarkModeSwitchEnabled=true
```
Comment (jonkoops) [inline themes/src/main/resources/theme/keycloak.v2/login/theme.properties:93]: For the regular theme properties we don't have to add any prefixes. In this case I'd say we can make this:
```suggestion
darkModeEnabled=true
```

File: themes/src/main/resources/theme/keycloak/welcome/index.ftl [hunk 3]
```diff
@@ -4,7 +4,27 @@
     <meta charset="utf-8">
     <meta name="robots" content="noindex, nofollow">
     <meta name="viewport" content="width=device-width, initial-scale=1">
+    <meta name="color-scheme" content="light dark">
```
Comment (jonkoops) [inline themes/src/main/resources/theme/keycloak/welcome/index.ftl:7]: Same here, this should be conditional based on whether the dark mode is enabled.

File: themes/src/main/resources/theme/keycloak/welcome/index.ftl [hunk 4]
```diff
@@ -4,7 +4,27 @@
...
     <title>Welcome to ${productName}</title>
+    <script type="module" async blocking="render">
```
Comment (jonkoops) [inline themes/src/main/resources/theme/keycloak/welcome/index.ftl:9]: Same here, this should be conditional based on whether the dark mode is enabled.

File: themes/src/main/resources/theme/keycloak/welcome/theme.properties [hunk 5]
```diff
@@ -5,3 +5,5 @@ styles=css/welcome.css
 
 # When set to true, the user will be redirected to the Administration Console if an administrative users already exists.
 redirectToAdmin=true
+kcDarkModeClass=pf-v5-theme-dark
+kcDarkModeSwitchEnabled=true
```
Comment (jonkoops) [inline themes/src/main/resources/theme/keycloak/welcome/theme.properties:9]: Same here, we can drop the prefix for this one.

```suggestion
darkModeEnabled=true
```

Comment (edewit) [reply themes/src/main/resources/theme/keycloak/welcome/theme.properties:9]: that means that this theme is wrong

Comment (edewit) [reply themes/src/main/resources/theme/keycloak/welcome/theme.properties:9]: also I could remove the `kcDarkModeClass` completely as all the classes are hardcoded anyway

Comment (jonkoops) [reply themes/src/main/resources/theme/keycloak/welcome/theme.properties:9]: Not really, there is a similar property right above it (`redirectToAdmin`). The main reason why the prefix exists is for the CSS class names, which can be a LOT.

File: themes/src/main/resources/theme/keycloak.v2/login/theme.properties (hunk 2 above)
Comment (ssilvert) [reply themes/src/main/resources/theme/keycloak.v2/login/theme.properties:93]: +1

Comment (ssilvert) [reply themes/src/main/resources/theme/keycloak.v2/login/theme.properties:93]: But wouldn't it all work and make code simpler if we just added this to the properties file?
```
# Comment out this line to disable dark mode
kcDarkModeClass=pf-v5-theme-dark
```
Then if `kcDarkModeClass` is null or empty then dark mode never gets added.

**Review by ssilvert — CHANGES_REQUESTED** [review]

File: themes/src/main/resources/theme/keycloak.v2/login/theme.properties (hunk 2 above)
Comment (edewit) [reply themes/src/main/resources/theme/keycloak.v2/login/theme.properties:93]: Yes that would also work, but @jonkoops wants the whole script removed when darkmode is not enabled

**Review by jonkoops — COMMENTED** [review]
As far as I am concerned this change is now looking good. I went ahead and added some details to the themes documentation as well as release notes, and aligned the properties so they are all named the same between themes.

@edewit @ssilvert @agagancarczyk WDYT?

**Review by edewit — DISMISSED** [review]

**Review by ssilvert — DISMISSED** [review]

**Review by agagancarczyk — DISMISSED** [review]

File: themes/src/main/resources/theme/keycloak.v2/login/theme.properties (hunk 2 above)
Comment (jonkoops) [reply themes/src/main/resources/theme/keycloak.v2/login/theme.properties]: Resolving this as the PR was approved by all.

**Comment by jonkoops** [comment]
Looks like the failures in the WebAuthN tests here are constant, re-basing this PR won't fix them :cry:

**Comment by jonkoops** [comment]
I am pretty confident this is the final incarnation. The tests should be working now. @edewit @ssilvert @agagancarczyk could we get this re-reviewed?

File: js/libs/ui-shared/src/utils/darkMode.ts [hunk 6]
```diff
@@ -1,19 +0,0 @@
...
-    classList.remove(DARK_MODE_CLASS);
-  }
-}
-
-export function initializeDarkMode() {
-  updateDarkMode(mediaQuery.matches);
-  mediaQuery.addEventListener("change", (event) =>
-    updateDarkMode(event.matches),
-  );
-}
```
Comment (edewit) [inline js/libs/ui-shared/src/utils/darkMode.ts:19]: No more dark theme for account and admin? Or you should add this to the index.ftl

**Review by edewit — CHANGES_REQUESTED** [review]

File: js/libs/ui-shared/src/utils/darkMode.ts (hunk 6 above)
Comment (jonkoops) [reply js/libs/ui-shared/src/utils/darkMode.ts:19]: Ahh, I forgot to push some local changes, should be good now.

**Review by edewit — APPROVED** [review]

**Review by ssilvert — APPROVED** [review]

**Pull request merged**